lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
   return TypeWidth >= TypeWidthThreshold || TypeDepth >= TypeDepthThreshold;
 }
 
+/// Check whether the type T1 is contained in the type T2
+static bool isContainedIn(CanType T1, CanType T2) {
+  return T2.findIf(
+      [&T1](Type T) -> bool { return T->getCanonicalType() == T1; });
+}
+
+/// Checks if a second substitution list is a expanded version of
+/// the first substitution list.
+/// This is the case if at least one of the substitution type in Subs2 is
+/// "bigger" than the corresponding substitution type in Subs1.
+/// Type T2 is "smaller" than type T1 if T2 is structurally contained in T1.
+static bool growingSubstitutions(SubstitutionList Subs1,
```

Comment:
@huonw I added support for you use-cases and used `TypeMatcher` for it. Please have a look.